### PR TITLE
KAFKA-8972 (2.4 blocker): clear all state for zombie task on TaskMigratedException

### PR DESCRIPTION
**Pull request by ableegoldman**
Third bugfix for the failing broker bounce system test with cooperative rebalancing:

**tl;dr** We need to remove everything associated with a task when it is closed, but in some cases (eg `AssignedTasks#commit`) on a `TaskMigratedException` we would close it as a zombie and then (only) remove the taskId from the `running` map. This left  its partitions, restorers, state stores, etc around and in an undefined state, causing exceptions when closing and/or opening the stores again.

**Longer explanation:**
In AssignedTasks (the abstract class from which the standby and active task variations extend) a commit failure (even due to broker down/unavailable) is treated as a TaskMigratedException after which the failed task is closed as a zombie and removed from running -- the remaining tasks (ie those still in running are then also closed as zombies in the subsequent onPartitionsLost

However we do not remove the closed task from runningByPartition nor do we remove the corresponding changelogs, if restoring, from the StoreChangelogReader since that applies only to active tasks, and AssignedTasks is generic/abstract. The changelog reader then retains a mapping from the closed task's changelog partition to its CompositeRestoreListener (and does not replace this when the new one comes along after the rebalance). The restore listener has a reference to a specific RocksDBStore instance, one which was closed when the task was closed as a zombie, so it accidentally tries to restore to the "old" RocksDBStore instance rather than the new one that was just opened.

Although technically this bug existed before KIP-429, it was only uncovered now that we remove tasks and clear their state/partitions/etc one at a time. We don't technically need to cherrypick the fix back earlier as before we just blindly clear all data structures entirely during an eager rebalance.

### CODE REVIEW

**Comment by ableegoldman** [comment]
@guozhangwang @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 1]
```diff
@@ -282,15 +274,19 @@ RuntimeException closeZombieTasks(final Set<TaskId> lostTasks, final List<TopicP
...
                 firstException.compareAndSet(null, closeNonRunning(true, created.get(id), lostTaskChangelogs));
             } else if (restoring.containsKey(id)) {
+                log.debug("Closing the restoring stream task {}.", id);
                 firstException.compareAndSet(null, closeRestoring(true, created.get(id), lostTaskChangelogs));
             } else if (running.containsKey(id)) {
+                log.debug("Closing the running stream task {}.", id);
                 firstException.compareAndSet(null, closeRunning(true, running.get(id), lostTaskChangelogs));
             } else {
-                // task may have already been closed as a zombie and removed from all task maps
+                log.warn("Skipping closing the lost stream task {} as it was already removed.", id);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Just some logging, but I left it in this PR instead of my other logging PR as it applies to this scenario (the warning here would have been helpful)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 2]
```diff
@@ -325,13 +321,7 @@ boolean maybeResumeSuspendedTask(final TaskId taskId,
...
                     // is triggered in the rebalance callback
                     log.info("Failed to resume {} {} since it got migrated to another thread already. " +
-                        "Closing it as zombie before triggering a new rebalance.", taskTypeName, task.id());
-                    final RuntimeException fatalException = closeZombieTask(task);
-                    running.remove(taskId);
-
-                    if (fatalException != null) {
-                        throw fatalException;
-                    }
+                             "Will trigger a new rebalance and close all tasks as zombies together.", taskTypeName, task.id());
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Found some other occurrences of closing a zombie and only partially clearing state, updated them all so that all tasks will be closed as zombies together during `onPartitionsLost`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 3]
```diff
@@ -384,6 +374,14 @@ void addToRestoring(final StreamTask task) {
         }
     }
 
+    private void removeTaskFromRestoring(final StreamTask task) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:377]: Some semi-related cleanup, to consolidate the state upkeep logic for all the various data structures in here

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 4]
```diff
@@ -269,8 +269,8 @@ void closeRevokedSuspendedTasks() {
 
         if (exception != null) {
             throw exception;
-        } else if (!assignedActiveTasks.isEmpty()) {
-            throw new IllegalStateException("TaskManager had leftover tasks after removing all zombies");
+        } else if (!(active.isEmpty() && assignedActiveTasks.isEmpty() && changelogReader.isEmpty())) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:272]: Tried to tighten up the safety check here by verifying that all the things we expect to be fully cleared after `onPartitionsLost` are indeed empty of state/old objects

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 5]
```diff
@@ -1425,7 +1425,7 @@ public void shouldNotSendOffsetsAndCommitTransactionNorStartNewTransactionOnSusp
     }
 
     @Test
-    public void shouldWrapProducerFencedExceptionWithTaskMigragedExceptionInSuspendWhenCommitting() {
+    public void shouldWrapProducerFencedExceptionWithTaskMigratedExceptionInSuspendWhenCommitting() {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:1432]: typo fix

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:377]: Should we loop over `task.changelogPartitions()` instead? For StreamTask the `partitions()` and `changelogPartitions()` should be non-overlapping except the shared source/changelog partitions.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: nit: `lost stream task` -> `zombie stream task`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 6]
```diff
@@ -69,6 +69,8 @@ public void register(final StateRestorer restorer) {
             stateRestorers.put(restorer.partition(), restorer);
 
             log.trace("Added restorer for changelog {}", restorer.partition());
+        } else {
+            log.trace("Already added restorer for changelog {}", restorer.partition());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: nit: Skip adding restorer ... since it already exist in the registered restorer. Also I think we can make this as DEBUG.

**Review by guozhangwang — COMMENTED** [review]
Made a pass on the PR.

**Comment by guozhangwang** [comment]
@ableegoldman could you rebase?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java [hunk 7]
```diff
@@ -303,22 +302,22 @@ public void shouldCommitRunningTasks() {
...
-        t1.close(false, true);
-        EasyMock.expectLastCall();
         EasyMock.replay(t1);
         addAndInitTask();
 
         try {
             assignedTasks.commit();
             fail("Should have thrown TaskMigratedException.");
-        } catch (final TaskMigratedException expected) { /* ignore */ }
+        } catch (final TaskMigratedException expected) {
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java]: This whole logic could be simplified as:
```
    private void verifyExceptionalState(ThrowingRunnable action) {
        assertThrows(TaskMigratedException.class, action);
        // This task should be closed as a zombie with all the other tasks during onPartitionsLost
        assertThat(assignedTasks.runningTaskIds(), equalTo(singleTaskId));

        EasyMock.verify(t1);
    }
```
so that new test just needs to put in the intended action. Here `singleTaskId` is a class level parameter I defined to replace the singleton list, which is not highly required.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java [hunk 8]
```diff
@@ -56,4 +56,8 @@
      */
     void clear();
 
+    /**
+     * @return whether the changelog reader has been cleared and is empty.
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java]: This comment is not very accurate, IMHO the changelog reader could also reside in an empty state when initialized.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 9]
```diff
@@ -392,8 +390,7 @@ int maybeCommitPerUserRequested() {
         int committed = 0;
         RuntimeException firstException = null;
 
-        for (final Iterator<StreamTask> it = running().iterator(); it.hasNext(); ) {
-            final StreamTask task = it.next();
+        for (final StreamTask task : running.values()) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:397]: Looks much better!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 10]
```diff
@@ -282,15 +274,19 @@ RuntimeException closeZombieTasks(final Set<TaskId> lostTasks, final List<TopicP
 
         for (final TaskId id : lostTasks) {
             if (suspended.containsKey(id)) {
+                log.debug("Closing the suspended stream task {}.", id);
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: I'm very suspicious about this logic here as it looks very easy for having one stream task resides in multiple maps (both restoring and suspended). Maybe we should put a to-do here to re-enforce an exclusive relation between task maps and throw exceptions when the expectation breaks? Right now it's very inefficient to debug around these different maps. One other suggestion is to make only one task map and only mark task state as `suspended, restoring, created, running`, but this potentially comes with performance penalty.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 11]
```diff
@@ -228,12 +224,8 @@ private RuntimeException closeNonRunning(final boolean isZombie,
     private RuntimeException closeRestoring(final boolean isZombie,
                                             final StreamTask task,
                                             final List<TopicPartition> closedTaskChangelogs) {
-        restoring.remove(task.id());
+        removeTaskFromRestoring(task);
         closedTaskChangelogs.addAll(task.changelogPartitions());
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:228]: Just to clarify, is there any ordering requirement between these two operations?
```
removeTaskFromRestoring(task);
closedTaskChangelogs.addAll(task.changelogPartitions());	     
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStandbyTasks.java [hunk 12]
```diff
@@ -53,7 +53,7 @@ int commit() {
         final List<TopicPartition> revokedChangelogs = new ArrayList<>();
         for (final Map.Entry<TaskId, Set<TopicPartition>> entry : revokedTasks.entrySet()) {
             final TaskId taskId = entry.getKey();
-            final Task task;
+            final StandbyTask task;
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStandbyTasks.java:66]: Is this necessary?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 13]
```diff
@@ -325,13 +321,7 @@ boolean maybeResumeSuspendedTask(final TaskId taskId,
                     // we need to catch migration exception internally since this function
                     // is triggered in the rebalance callback
                     log.info("Failed to resume {} {} since it got migrated to another thread already. " +
-                        "Closing it as zombie before triggering a new rebalance.", taskTypeName, task.id());
-                    final RuntimeException fatalException = closeZombieTask(task);
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:329]: The new diff contains two `closeZombieTask` usages, where one of them in AssignedStreamTasks has
```
 firstException.compareAndSet(null, closeZombieTask(task));
 prevActiveTasks.remove(id);
```
which makes me wonder whether standby tasks also have to update some internal state like `closeRevokedStandbyTasks`, which makes me wonder should we disgard this function, as using it alone is usually not good enough and may encourage bugs since it gives people impression it could do more than just closing a task.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 4 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:272]: Should we also check other maps in `TaskManager`, like `assignedStandbyTasks`or `revokedActiveTasks`. Again just my feeling about the current setup, I don't find it intuitive to deal with multiple mapping structs and they look so similar to me :) It's very easy to forget what they refer to really soon.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 3 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:377]: We should also consider throwing exception here if a task spams multiple structs

**Review by abbccdda — COMMENTED** [review]
Thanks, left some comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStandbyTasks.java (hunk 12 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStandbyTasks.java:66]: Yeah actually `removeTaskFromRunning` takes in a `<T extends Task>` type specifically. But it's good cleanup anyway

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 11 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:228]: Nope, `removeTaskFromRestoring` just removes the task from `AssignedStreamTasks` -- it shouldn't affect the task or its fields in any way

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Ack

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 3 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:377]: @guozhangwang Good catch, we should actually loop over both of them since that's what we do in `addTaskToRestoring` above
@abbccdda I do agree with you about more strongly enforcing task state but I'm not sure this would be the most practical/appropriate way to do it. See other comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 10 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Yeah I was thinking along the same lines, and in particular that we should consider enforcing not only that tasks be in at most 1 map, but _exactly_ one (until we start closing them) -- I actually started doing some more significant refactoring in this PR but ended up splitting it out for a follow-up PR. I figured this one should just contain the minimal bugfix and avoid accidentally breaking things at the last minute.
WDYT?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java]: Hm, true. I was struggling to phrase this because technically it doesn't make sense for the `ChangelogReader` interface to be "empty" since it's not a data structure -- the specific `StoreChangelogReader` implementation just happens to keep (a lot of) state.
How about "...reader has just been cleared or is uninitialized"?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java]: Yeah I've been thinking about what the "right" thing to do here is and have some actual changes in mind but I pulled them out into a separate follow-up PR to keep this one tight. But, doesn't hurt to make debug here

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:272]: Well, we actually don't clear the standbys on `onPartitionsLost` as they weren't part of the group subscription to begin with. `revokedActiveTasks` and `addedActiveTasks` just tell us how the assignment changed since the previous one, and don't matter at all to us here (they aren't required to be empty, since they're not updated except after a rebalance)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java]: Ack

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 13 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java:329]: Good question, but `StandbyTask#close` explicitly clarifies it can never be a zombie. With the removal of most occurrences of `closeZombieTask` from `AssignedTasks` there's actually a good amount of further cleanup to do here, especially around the general `close`
That's actually specifically what I started doing before I realized the changes were getting out of hand for this simple bugfix PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java [hunk 14]
```diff
@@ -571,5 +539,16 @@ private void mockRunningTaskSuspension() {
...
+        // This action is expected to throw a TaskMigratedException
+        assertThrows(TaskMigratedException.class, action);
+
+        // This task should be closed as a zombie with all the other tasks during onPartitionsLost
+        assertThat(assignedTasks.runningTaskIds(), equalTo(expectedRunningTaskIds));
+
+        EasyMock.verify(t1);
+    }
 
 }
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java:554]: nit: empty line

**Review by abbccdda — APPROVED** [review]
LGTM after reviewing added comments

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java (hunk 14 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java:554]: You mean right before the end of the file? A lot of them seem to have that, not sure if intentional or not though

**Comment by ableegoldman** [comment]
Java 8 and Java 11/2.12 passed, Java 11/2.13 failed with unrelated flaky 
`kafka.admin.ReassignPartitionsClusterTest.shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup`

retest this please

**Comment by guozhangwang** [comment]
Resolved minor conflicts and run `./gradlew unitTest` before cherry-picked to 2.4. cc @omkreddy
